File: Python/PE005.py
"""
Problem 5: Smallest multiple
URL: https://projecteuler.net/problem=5

    2520 is the smallest number that can be divided by each of the
    numbers from 1 to 10 without any remainder.

    What is the smallest positive number that is evenly divisible
    (divisible with no remainder) by all of the numbers from 1 to 20?
"""
import math
import itertools


######################################################################
# Constants
######################################################################

FIRST = 1
LAST = 20


######################################################################
# Functions
######################################################################

def find_number(first, last):
    """
    Find the smallest natural divisible by a range of numbers.
    :param first: The first value of the divisors range.
    :param last: The last value of the divisors range.
    :return: The smallest natural number.
    """
    # First, get all the combinations of the divisors:
    divisors = range(first, last + 1)
    combinations = set()
    for size in range(1, len(divisors)):
        values = itertools.combinations(divisors, size)
        combinations |= set(math.prod(value) for value in values)
    # Then check each combination in ascending order, to
    # find the final result, and return it when found:
    for victim in sorted(combinations):
        success = True
        for divisor in divisors:
            if (victim % divisor) != 0:
                success = False
                break

        if success:
            return victim
    # Otherwise return None:
    return None


def main():
    """
    Main entry for the problem solver.
    """
    result = find_number(FIRST, LAST)

    # Show the final result of the problem:
    message = "The smallest positive number that is a multiple "
    message += f"of all numbers from 1 to 20 is {result}."
    print(message)


######################################################################
# Functions (Slow version)
######################################################################

def find_number_slow(first, last):
    """
    Find the smallest natural divisible by a range of numbers.
    :param first: The first value of the divisors range.
    :param last: The last value of the divisors range.
    :return: The smallest natural number.
    """
    result = 1
    while True:
        # Check if the current number is divisible:
        success = True
        for divisor in range(first, last + 1):
            if (result % divisor) != 0:
                success = False
                break
        # Return the result when success, otherwise
        # select the next candidate:
        if success:
            return result
        else:
            result += 1


######################################################################
# Module execution
######################################################################

if __name__ == "__main__":
    main()

File: Erlang/src/pe005.erl
%%%======================================================================
%%% @author Gorka Suárez García
%%% @copyright (C) 2022-2023, Gorka Suárez García
%%% @doc
%%% 2520 is the smallest number that can be divided by each of the
%%% numbers from 1 to 10 without any remainder.
%%%
%%% What is the smallest positive number that is evenly divisible
%%% (divisible with no remainder) by all of the numbers from 1 to 20?
%%% @end
%%%======================================================================
-module(pe005).
-author("Gorka Suárez García").
-export([main/0, result/0]).

-define(FIRST, 1).
-define(LAST, 20).

%%-----------------------------------------------------------------------
%% @doc
%% Main entry for the problem solver.
%% @end
%%-----------------------------------------------------------------------
main() ->
    io:format("The smallest positive number that is a multiple of all "
              "numbers from 1 to 20 is ~p.~n", [result()]).

%%-----------------------------------------------------------------------
%% @doc
%% Main result for the problem solver.
%% @end
%%-----------------------------------------------------------------------
result() ->
    {_, Result} = find_number(?FIRST, ?LAST),
    Result.

%%-----------------------------------------------------------------------
%% @private
%% @doc
%% Find the smallest natural divisible by a range of numbers.
%% @param First The first value of the divisors range.
%% @param Last The last value of the divisors range.
%% @returns The smallest natural number.
%% @end
%%-----------------------------------------------------------------------
find_number(First, Last) ->
    Numbers = lists:seq(First, Last),
    tools:find(
        lists:seq(1, length(Numbers)),
        fun(Size) ->
            tools:find(
                tools:combinations(Numbers, Size),
                fun(Combination) ->
                    X = tools:product(Combination),
                    case lists:all(fun(Y) -> (X rem Y) == 0 end, Numbers) of
                        true -> {ok, X};
                        _ -> continue
                    end
                end
            )
        end
    ).
